Expand and document PSRT coordination process (#1769)
python/devguide: developer-workflow/psrt.rst

Triaging a vulnerability report
-------------------------------

PSRT members coordinate reports from when they are first submitted
to a "finished" state. Finished states include
marking a report as a "non-issue", opening a public issue on GitHub,
or a merged patch with an accompanying CVE and advisory to
``security-announce@python.org``. Reports should reach a finished
state within 90 days of being received by the PSRT.

Reports enter the system through ``security@python.org`` or
on a project GitHub Security Advisory (GHSA) ticketing system.
For projects that use GHSA, reports to ``security@python.org``
should have reporters `re-open their report using GHSA`_.

.. raw:: html
   :file: ../_static/psrt-coordinator-report.svg

.. _re-open their report using GHSA: #submit-using-github-security-advisories

New report in GitHub Security Advisories (GHSA)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Once a report is in GHSA, a "Coordinator" must be assigned
to be responsible for moving the report through the process.
The "Coordinator" role is assigned using a "Credit" in a GHSA ticket
(Select 'Edit' > 'Credit' > Add GitHub username and the role 'Coordinator').

> If a GHSA ticket is idle for 3 days without a coordinator

Reviewer: ```suggestion
If a GHSA ticket is idle for three days without a coordinator
```

assigned a non-"Release Manager" / "Steering Council"

> PSRT member will be automatically assigned as coordinator by the PSRT bot.

Reviewer 2: ```suggestion
assigned a PSRT member who is not a Release Manager
or Steering Council member will be automatically assigned
as coordinator by the PSRT bot.
```
A little suggestion for rewording.
I'd also recommended a note as to how the selection [will] takes place. Is it just `random.choice(psrt-members)`? Does it take into account whether the person is already coordinating another report?

Author: My thought is it would account for folks who are already coordinating reports to "smooth" out the load where possible, I don't think that would be controversial?

If a coordinator can't complete the process

> they must find a replacement coordinator in the PSR

Reviewer 2: ```suggestion
they must find a replacement coordinator in the PSRT
```

and re-assign the GHSA ticket.

Determining whether a report is a vulnerability
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. raw:: html
   :file: ../_static/psrt-coordinator-triage.svg

The coordinator will make a determination about a report, either
marking the ticket as a non-issue, as an issue that isn't a security
vulnerability, or as a security vulnerability. If the Coordinator needs

> help from core team member experts in making the determination, the

Reviewer: ```suggestion
help from core team experts in making the determination, the
```

the Coordinator must prepare before sending an advisory and
closing the GHSA ticket:

* Severity calculated using CVSSv4.
* Pull request containing the fix merged with a public GitHub issue.
* Advisory title and short description of the vulnerability.

Severity scoring
~~~~~~~~~~~~~~~~

Severity of a vulnerability can be difficult to assess
objectively due to not knowing how software is used
in all situations. Severity is calculated from expected
or known use, not from worst-case hypothetical scenarios.

The PSRT and PSF CNA use `CVSSv4`_ for calculating
the severity of a vulnerability. GHSA tickets provide a
CVSSv4 calculator within the ticket UI. Note that GitHub defaults to CVSSv3,
change the scoring algorithm to CVSSv4 before scoring in the ticket UI.
As with all aspects of PSRT operations, coordinators are encouraged to ask
for help in calculating a severity from other PSRT members.

.. _CVSSv4: https://www.first.org/cvss/v4.0/

Developing a patch privately
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Patch development can initially be done privately by selecting the
'Start a temporary private fork' button within the GHSA ticket. Note that

> due to the size of git repositories, this fork repository may

Reviewer: ```suggestion
due to the size of Git repositories, this fork repository may
```

take several minutes to create. Once the fork has been created any PSRT member
or GHSA collaborator can clone the fork and develop a fix and push a branch:

.. code-block:: shell

    git clone https://github.com/python/cpython-ghsa-abcd-efgh-ijkl.git
    cd cpython-ghsa-abcd-efgh-ijkl
    git checkout -b advisory-fix-1

    # (develop a fix)

    git add ...
    git commit -m "Fix"
    git push origin advisory-fix-1

From here a pull request can be opened within the GHSA ticket
to be reviewed privately. The coordinator can add core team experts
to help with review if reviewers aren't available within PSRT.
Once the patch has been approved
a public GitHub issue and pull request can be opened.
Generate a blurb in the category ``Security`` for the public pull request.
Public issues and pull requests for security fixes should use the

> ``type-security`` label.

Reviewer: ```suggestion
:gh-label:`type-security` label.
```

        --output ghsa-abcd-efgh-ijkl.patch

This patch can then be applied and pushed to the public GitHub repository:

.. code-block:: shell

    git remote -v
    origin    ssh://git@github.com/.../cpython (fetch)
    origin    ssh://git@github.com/.../cpython (fetch)

    git checkout -b branch-name
    git apply ./ghsa-abcd-efgh-ijkl.patch
    git push origin branch-name

.. warning:: **IMPORTANT:** CPython's backport infrastructure
    is used for tracking backported patches. Use **one GitHub issue
    per CVE** to accurately track backports of vulnerability fixes.
    For new CVEs, even when related to a previous issue, **open a
    new GitHub issue** to accurately track fixed versions.

.. warning:: **IMPORTANT:** Don't select the green 'Merge pull request'
    or 'Publish advisory' buttons within GHSA. Advisories are published
    to the mailing list, and the 'Merge pull request' button within
    GHSA bypasses all continuous integration and branch protection
    steps. Use a public pull request instead.

Publishing an advisory
~~~~~~~~~~~~~~~~~~~~~~

Once the vulnerability fix has been merged in a public GitHub into the ``main``

> branch, an advisory must be published. The advisory requires the severity,

Reviewer 2: ```suggestion
Once the vulnerability fix has been merged into the main branch via a
public GitHub pull request, an advisory must be published.
The advisory requires the severity,
```
I found "Once the vulnerability fix has been merged in a public GitHub" a little difficult to understand, maybe something like this?

> a title, and a short description of the vulnerable module, function(s),

Reviewer: ```suggestion
a title, and a short description of the vulnerable module, functions,
```

behavior and fix. This short description can optionally include mitigation steps
if applying the patch isn't the only way to mitigate the vulnerability.

* Send an email to the ``security-announce@python.org`` mailing list
  using the `advisory template`_, including title, severity, description.
* The advisory email will be received by PSF CVE Numbering Authority
  operators and used to publish a CVE record.
* Begin the backporting process for all Python branches still receiving

>   security updates. Add the ``type-security`` and ``release-blocker`` labels

Reviewer: ```suggestion
  security updates. Add the :gh-label:`type-security` and :gh-label:`release-blocker` labels
```

  to each backport pull request so that release managers can find them prior

>   to releases.

Reviewer: ```suggestion
  to releasing.
```

* Running nomination elections, including counting final votes and giving
  the Steering Council an opportunity to veto nominations via email.

Triaging a vulnerability report
-------------------------------

PSRT members coordinate reports from when they are first submitted
to a "finished" state. Finished states include
marking a report as a "non-issue", opening a public issue on GitHub,
or a merged patch with an accompanying CVE and advisory to
``security-announce@python.org``. Reports should reach a finished
state within 90 days of being received by the PSRT.

Reports enter the system through ``security@python.org`` or
on a project GitHub Security Advisory (GHSA) ticketing system.
For projects that use GHSA, reports to ``security@python.org``

> should have reporters `re-open their report using GHSA`_.

Reviewer: ```suggestion
should have reporters :ref:`re-open their report using GHSA <submit-using-ghsa>`.
```

.. raw:: html
   :file: ../_static/psrt-coordinator-report.html

.. _re-open their report using GHSA: #submit-using-github-security-advisories

> 

Reviewer: ```suggestion
```

for help in calculating a severity from other PSRT members.

.. _CVSSv4: https://www.first.org/cvss/v4.0/

Developing a patch privately
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Patch development can initially be done privately by selecting the
'Start a temporary private fork' button within the GHSA ticket. Note that
due to the size of Git repositories, this fork repository may
take several minutes to create. Once the fork has been created any PSRT member
or GHSA collaborator can clone the fork and develop a fix and push a branch:

.. code-block:: shell

    git clone https://github.com/python/cpython-ghsa-abcd-efgh-ijkl.git
    cd cpython-ghsa-abcd-efgh-ijkl
    git checkout -b advisory-fix-1

    # (develop a fix)

    git add ...
    git commit -m "Fix"
    git push origin advisory-fix-1

From here a pull request can be opened within the GHSA ticket
to be reviewed privately. The coordinator can add core team experts
as collaborators as necessary to the GHSA to develop the patch and to review
the patch if the expertise for patch development or review
aren't available within PSRT. Coordinators are not necessarily responsible

> with developing patches themselves: only with seeing that a patch is

Reviewer: ```suggestion
for developing patches themselves: only with seeing that a patch is
```

in various points in the process above. They are not required to be sent as-is,
please feel free to adapt them as needed for the current context.

> Submit using GitHub Security Advisories
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: Better to link with a ref than with an anchor, as anchors change if we adjust headers later:

```suggestion
.. _submit-using-ghsa:

Submit using GitHub Security Advisories
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

   If you are interested in working on this further, you can optionally open a
   public issue on GitHub.

Advisory email
~~~~~~~~~~~~~~

::

   Title: [{CVE-YYYY-XXXX}] {title}

   There is a {LOW, MEDIUM, HIGH, CRITICAL} severity vulnerability affecting {project}.

   {description}

   Please see the linked CVE ID for the latest information on affected versions:

   * https://www.cve.org/CVERecord?id={CVE-YYYY-XXXX}

>    * {pull request URL}

Reviewer 2: Is it better to link the pull request or the commit? The commit is a simpler/direct view of the change, and does have a link to both the issue and PR.

Author: Linking to the commit works, I chose PR because the PSRT bot fills in commits on the CVE record automatically from the issue.

Reviewer: I think PR is a bit more useful, there's more context: discussion, links to issue, and commit.
